src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+            else
+            {
+                return new EditorCompletion.FilteredCompletionModel(highlightedList, 0, updatedFilters, EditorCompletion.UpdateSelectionHint.SoftSelected, centerSelection: true, uniqueItem: null);
+            }
+        }
+
+        private bool IsBetterDeletionMatch(FilterResult result1, FilterResult result2)
+        {
+            var item1 = result1.CompletionItem;
+            var item2 = result2.CompletionItem;
+
+            var prefixLength1 = item1.FilterText.GetCaseInsensitivePrefixLength(result1.FilterText);
+            var prefixLength2 = item2.FilterText.GetCaseInsensitivePrefixLength(result2.FilterText);
+
+            // Prefer the item that matches a longer prefix of the filter text.
+            if (prefixLength1 > prefixLength2)
+            {
+                return true;
+            }
+
+            if (prefixLength1 == prefixLength2)
+            {
+                // If the lengths are the same, prefer the one with the higher match priority.
+                // But only if it's an item that would have been hard selected.  We don't want
+                // to aggressively select an item that was only going to be softly offered.
+
+                var item1Priority = CalculatePriority(item1);
+                var item2Priority = CalculatePriority(item2);
+
+                if (item1Priority > item2Priority)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private int CalculatePriority(EditorCompletion.CompletionItem item)
+            => GetOrCreateRoslynItem(item)?.Rules?.SelectionBehavior != CompletionItemSelectionBehavior.HardSelection
+                    ? MatchPriority.Default
+                    : GetMatchPriority(item);
+
+        private EditorCompletion.FilteredCompletionModel HandleNormalFiltering(
+            ImmutableArray<EditorCompletion.CompletionItem> sortedList,
+            ITextSnapshot snapshot,
+            Document document,
+            int? caretPosition,
+            string filterText,
+            ImmutableArray<EditorCompletion.CompletionFilterWithState> filters,
+            EditorCompletion.CompletionTriggerReason triggerReason,
+            List<FilterResult> itemsInList,
+            bool mustSetSelection)
+        {
+            var completionService = document.GetLanguageService<CompletionService>();
+            if (completionService == null)
+            {
+                var listWithSelections = GetHighlightedList(itemsInList, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, filters, EditorCompletion.UpdateSelectionHint.NoChange, centerSelection: true, uniqueItem: default);
+            }
+
+            var matchingItems = itemsInList
+                .Where(r => r.MatchedFilterText)
+                .Select(t => GetOrCreateRoslynItem(t.CompletionItem))
+                .AsImmutable();
```

Comment:
codestyle was changed here as well.  why?  this makes matching everything up again and reviewing the change super difficult.